### ISSUE TITLE
Compatibility with raspberry pi camera module 3?

### ISSUE THREAD

**Issue by biprav** [silvanmelchior/RPi_Cam_Web_Interface#692]
Hi.
I just bought a 12 MP raspberry pi camera module 3 and have connected it with a Pi zero 2W. I have tried using the cam web interface with the camera module 3 but it gives various errors. Can you kindly tell me whether it works with camera module 3?

Thank you.

**Comment by roberttidey** [comment]
This software uses the GPU legacy camera support which only supports v1 and v2 module rapsberry cameras.

**Comment by biprav** [comment]
> This software uses the GPU legacy camera support which only supports v1 and v2 module rapsberry cameras.

Okay. Thank you for responding.

**Comment by ways** [comment]
No workaround to pipe video from camera v3 to this?

**Comment by kaihokori** [comment]
Hi @biprav, you might be interested in [this](https://github.com/silvanmelchior/RPi_Cam_Web_Interface/issues/688#issuecomment-2437456955). This program should be usable on the pi camera module 3 as it uses Picamera2 (which is developed by Raspberry Pi themselves) as a foundation. Be aware that not all functionality has been implemented but it should be a good starting point ^_^

**Comment by kkaazzuukkii4** [comment]
Yes — Camera Module 3 works today. Posting a hardware-verified report, since I couldn't find one in this thread.

The stock `raspimjpeg` backend will never support it (MMAL-based; IMX708 only exists under libcamera). But you can replace just the backend with [RasPyCam](https://github.com/Windermere-Technology/RasPyCam) and leave this project's web interface **completely unmodified** — the FIFO command interface is compatible.

## Tested on

| | |
|---|---|
| Board | Raspberry Pi 4 Model B Rev 1.4 |
| OS | Raspberry Pi OS Bookworm (Debian 12), 64-bit |
| Camera | Camera Module 3 Wide NoIR — `imx708_wide_noir` |
| libcamera | 0.5.2 (rpicam-apps 1.9.0) |
| Backend | RasPyCam `main` + its open PRs [29](https://github.com/Windermere-Technology/RasPyCam/pull/29) and [26](https://github.com/Windermere-Technology/RasPyCam/pull/26) |
| Frontend | this project, **unmodified** |

## Working

| Feature | Result |
|---|---|
| Preview stream | ✅ |
| Still capture | ✅ 4608×2592 |
| Video recording | ✅ H.264 1080p30, valid MP4 |
| Motion detection | ✅ triggers recording |
| Timelapse | ✅ |
| Image controls (`br` `co` `sh` `wb` `ss` `is` `ec`) | ✅ |
| Web UI | ✅ no changes needed |

## Setup

No fork needed; the two open RasPyCam PRs can be pulled straight from upstream.

```bash
# Confirm the OS sees the camera BEFORE installing anything
rpicam-hello --list-cameras     # must list imx708

git clone https://github.com/Windermere-Technology/RasPyCam.git
cd RasPyCam && git checkout -b cm3
git fetch origin pull/29/head:pr29 && git merge --no-edit pr29   # Debian 13 support + 'im' fix
git fetch origin pull/26/head:pr26 && git merge --no-edit pr26   # nginx + PHP 8.2 socket path

sudo apt install -y python3-opencv
./install.sh
```

## Two things that will trip you up

**Autostart is broken on Bookworm — the camera won't come back after a reboot.** The installer appends to `/etc/rc.local`, which doesn't exist on Bookworm, so the file is created without a shebang and systemd can't execute it (`status=203/EXEC`). Fix:

```bash
sudo sed -i '1i #!/bin/sh' /etc/rc.local
```

Easy to miss, because running `sudo /etc/rc.local` by hand *succeeds* — the invoking shell falls back to interpreting it. Test with `systemctl start rc-local` instead. Verified fixed through a real reboot.

**`vcgencmd get_camera` lies.** It still reports `detected=0` on Bookworm with a perfectly working camera, because it describes the legacy stack that no longer exists. Use `rpicam-hello --list-cameras`.

## Main limitation

**No autofocus control** — the headline feature of Camera Module 3. Neither the FIFO command set nor the web UI exposes focus. I've implemented a `fo` command locally (Manual / Auto / Continuous + lens position) and confirmed on hardware that the lens moves; I plan to submit it to RasPyCam.

Also worth knowing: with Trixie now current, Raspberry Pi Imager's **"Raspberry Pi OS (Legacy, 64-bit)" is Bookworm**, which is what I used. The label used to mean Bullseye.

I hit a few other things (MP4Box/gpac, the schedule page timezone, resource usage) that belong in their own threads rather than here — I'll post those separately.

@roberttidey — thanks for the clear explanation earlier in this thread of *why* the MMAL pipeline couldn't simply be ported. That saved me from going down the wrong path, and everything I found is consistent with it: the stock backend genuinely can't support this camera. Flagging this in case it's useful to point future reports here. Thanks also to the RasPyCam team for doing the hard part — this is just confirmation that their work handles Camera Module 3.

---

*Disclosure: I used Claude (Anthropic) to help investigate and write up this report. All testing was run on my own hardware and the results above are real command output, not generated text. Happy to provide any raw logs on request.*
